src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -51,287 +53,240 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
+
+type PhaseInfo = { Phase: Phase; Func: PhaseFunc }
+
+type PhaseInfos = PhaseInfo[]
+
+[<RequireQualifiedAccess>]
+module private ParallelOptimization =
+    open Optimizer
 
-    type Node =
+    /// Identifies a work item scheduled independently of others - consists of a (file) index and OptimizationPhase.
+    /// There are (NumberOfFiles * NumberOfPhases) nodes in the whole optimization process.
+    type private Node =
         {
-            Idx: int
-            Phase: OptimizationPhase
+            FileIdx: int
+            Phase: PhaseIdx
         }
 
-        override this.ToString() = $"[{this.Idx}-{this.Phase}]"
+        override this.ToString() = $"[{this.FileIdx}-{this.Phase}]"
 
-    let collectResults
-        (inputs: FileResultsComplete[])
+    /// Final processing of file results to produce output needed for further compilation steps.
+    let private collectFinalResults
+        (fileResults: PhaseRes[])
         : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
-        let files =
-            inputs
-            |> Array.map
-                (fun {
-                         Phase1 = phase1
-                         Phase2 = _phase2
-                         Phase3 = phase3
-                     } ->
-                    let (_, _, implFileOptData, _), optimizeDuringCodeGen = phase1
-                    let _, implFile = phase3
-
-                    let implFile =
-                        {
-                            ImplFile = implFile
-                            OptimizeDuringCodeGen = optimizeDuringCodeGen
-                        }
-
-                    implFile, implFileOptData)
+        let finalFileResults =
+            fileResults
+            |> Array.map (fun (file, res) ->
+                let implFile =
+                    {
+                        ImplFile = file
+                        OptimizeDuringCodeGen = res.OptDuringCodeGen
+                    }
 
-        let lastFilePhase1Env =
-            inputs
-            |> Array.last
-            |> fun { Phase1 = phase1 } ->
-                let (optEnvPhase1, _, _, _), _ = phase1
-                optEnvPhase1
+                implFile, res.OptInfo)
 
-        files, lastFilePhase1Env
+        let lastFileFirstLoopEnv =
+            fileResults |> Array.last |> (fun (_file, res) -> res.OptEnvFirstLoop)
 
-    let getPhase2Res (p: FileResults) = p.Phase2 |> Option.get
+        finalFileResults, lastFileFirstLoopEnv
 
-    let getPhase3Res (p: FileResults) =
-        p.Phase3 |> Option.get |> (fun (env, _) -> env)
+    let private raiseNoResultsExn (node: Node) =
+        raise (exn $"Unexpected lack of results for {node}")
 
-    let optimizeFilesInParallel
+    let optimizeFilesInParallel2
         (env0: IncrementalOptimizationEnv)
-        ((phase1, phase2, phase3): FilePhaseFuncs)
+        (phases: PhaseInfos)
         (files: CheckedImplFile list)
```

Comment:
You are indexing into 'files' and also gettings its .Length -> array would be more appropriate over list.

Reply:
Will do.